File: apps/api/app/order_store.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

OrderStatus = Literal["open", "confirmed"]


@dataclass
class OrderItem:
    product_id: str
    name: str
    price: float
    quantity: int


@dataclass
class Order:
    items: list[OrderItem] = field(default_factory=list)
    status: OrderStatus = "open"

    @property
    def total(self) -> float:
        return sum(item.price * item.quantity for item in self.items)

    def to_dict(self) -> dict:
        return {
            "status": self.status,
            "items": [
                {
                    "product_id": item.product_id,
                    "name": item.name,
                    "price": item.price,
                    "quantity": item.quantity,
                    "subtotal": round(item.price * item.quantity, 2),
                }
                for item in self.items
            ],
            "total": round(self.total, 2),
        }


class OrderStore:
    def __init__(self) -> None:
        self._order = Order()

    def reset(self) -> None:
        self._order = Order()

    def snapshot(self) -> dict:
        return self._order.to_dict()

    def add_item(self, product_id: str, name: str, price: float, quantity: int = 1) -> dict:
        if self._order.status == "confirmed":
            return {"error": "Order is already confirmed."}

        for item in self._order.items:
            if item.product_id == product_id:
                item.quantity += quantity
                return {"success": True, "order": self.snapshot()}

        self._order.items.append(
            OrderItem(product_id=product_id, name=name, price=price, quantity=quantity)
        )
        return {"success": True, "order": self.snapshot()}

    def remove_item(self, product_id: str, quantity: int | None = None) -> dict:
        if self._order.status == "confirmed":
            return {"error": "Order is already confirmed."}

        for index, item in enumerate(self._order.items):
            if item.product_id == product_id:
                if quantity is None or quantity >= item.quantity:
                    self._order.items.pop(index)
                else:
                    item.quantity -= quantity
                return {"success": True, "order": self.snapshot()}

        return {"error": f"Product '{product_id}' not found in order."}

    def confirm(self) -> dict:
        if not self._order.items:
            return {"error": "Cannot confirm an empty order."}

        self._order.status = "confirmed"
        return {"success": True, "order": self.snapshot()}

File: apps/server/src/services/order-store.ts
type OrderStatus = "open" | "confirmed";

interface OrderItem {
  product_id: string;
  name: string;
  price: number;
  quantity: number;
  image_url: string;
}

interface Order {
  items: OrderItem[];
  status: OrderStatus;
  customer_name: string | null;
}

export class OrderStore {
  private order: Order = { items: [], status: "open", customer_name: null };

  reset(): void {
    this.order = { items: [], status: "open", customer_name: null };
  }

  snapshot(): Record<string, unknown> {
    const payload: Record<string, unknown> = {
      status: this.order.status,
      items: this.order.items.map((item) => ({
        product_id: item.product_id,
        name: item.name,
        price: item.price,
        quantity: item.quantity,
        subtotal: Math.round(item.price * item.quantity * 100) / 100,
        image_url: item.image_url,
      })),
      total: Math.round(this.total * 100) / 100,
    };
    if (this.order.customer_name) {
      payload.customer_name = this.order.customer_name;
    }
    return payload;
  }

  get total(): number {
    return this.order.items.reduce((sum, item) => sum + item.price * item.quantity, 0);
  }

  loadSnapshot(payload: Record<string, unknown>): void {
    const items: OrderItem[] = [];
    for (const raw of (payload.items as unknown[]) ?? []) {
      if (!raw || typeof raw !== "object") continue;
      const item = raw as Record<string, unknown>;
      items.push({
        product_id: String(item.product_id ?? ""),
        name: String(item.name ?? ""),
        price: Number(item.price ?? 0),
        quantity: Number(item.quantity ?? 0),
        image_url: String(item.image_url ?? ""),
      });
    }

    let status = String(payload.status ?? "open") as OrderStatus;
    if (status !== "open" && status !== "confirmed") status = "open";

    const customerName = payload.customer_name;
    this.order = {
      items,
      status,
      customer_name:
        customerName != null && String(customerName).trim()
          ? String(customerName).trim()
          : null,
    };
  }

  addItem(
    productId: string,
    name: string,
    price: number,
    quantity = 1,
    imageUrl = "",
  ): Record<string, unknown> {
    if (this.order.status === "confirmed") {
      return { error: "Order is already confirmed." };
    }

    for (const item of this.order.items) {
      if (item.product_id === productId) {
        item.quantity += quantity;
        if (imageUrl && !item.image_url) item.image_url = imageUrl;
        return { success: true, order: this.snapshot() };
      }
    }

    this.order.items.push({
      product_id: productId,
      name,
      price,
      quantity,
      image_url: imageUrl,
    });
    return { success: true, order: this.snapshot() };
  }

  removeItem(productId: string, quantity?: number): Record<string, unknown> {
    if (this.order.status === "confirmed") {
      return { error: "Order is already confirmed." };
    }

    const index = this.order.items.findIndex((item) => item.product_id === productId);
    if (index === -1) {
      return { error: `Product '${productId}' not found in order.` };
    }

    const item = this.order.items[index]!;
    if (quantity === undefined || quantity >= item.quantity) {
      this.order.items.splice(index, 1);
    } else {
      item.quantity -= quantity;
    }
    return { success: true, order: this.snapshot() };
  }

  cancelOrder(): Record<string, unknown> {
    if (this.order.status === "confirmed") {
      return { error: "Order is already confirmed." };
    }
    this.order.items = [];
    this.order.customer_name = null;
    return { success: true, order: this.snapshot() };
  }

  setCustomerName(name: string): Record<string, unknown> {
    const trimmed = name.trim();
    if (!trimmed) return { error: "Customer name cannot be empty." };
    this.order.customer_name = trimmed;
    return { success: true, order: this.snapshot() };
  }

  confirm(): Record<string, unknown> {
    if (!this.order.items.length) {
      return { error: "Cannot confirm an empty order." };
    }
    this.order.status = "confirmed";
    return { success: true, order: this.snapshot() };
  }
}
